flink-runtime/src/main/java/org/apache/flink/runtime/jobmaster/JobMaster.java

```diff
@@ -1113,25 +1113,23 @@ private void assignExecutionGraph(
 
 	private void resetAndScheduleExecutionGraph() throws Exception {
 		validateRunsInMainThread();
-
-		final CompletableFuture<Void> executionGraphAssignedFuture;
-
+		MainThreadExecutor mainThreadExecutor = getMainThreadExecutor();
 		if (executionGraph.getState() == JobStatus.CREATED) {
-			executionGraphAssignedFuture = CompletableFuture.completedFuture(null);
+			scheduleExecutionGraph();
 		} else {
 			suspendAndClearExecutionGraphFields(new FlinkException("ExecutionGraph is being reset in order to be rescheduled."));
 			final JobManagerJobMetricGroup newJobManagerJobMetricGroup = jobMetricGroupFactory.create(jobGraph);
 			final ExecutionGraph newExecutionGraph = createAndRestoreExecutionGraph(newJobManagerJobMetricGroup);
 
-			executionGraphAssignedFuture = executionGraph.getTerminationFuture().handleAsync(
-				(JobStatus ignored, Throwable throwable) -> {
-					assignExecutionGraph(newExecutionGraph, newJobManagerJobMetricGroup);
-					return null;
-				},
-				getMainThreadExecutor());
+			executionGraph.getTerminationFuture()
+				.handleAsync(
+					(JobStatus ignored, Throwable throwable) -> {
+						assignExecutionGraph(newExecutionGraph, newJobManagerJobMetricGroup);
+						return null;
+					},
+					mainThreadExecutor)
```

Comment:
`getTerminationFuture` should always be completed from the main thread and, thus, there should be no need for `mainThreadExecutor` here.